src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Runtime;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
 using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.Optimizations;
+using Microsoft.Quantum.QsCompiler.SymbolManagement;
 using Microsoft.Quantum.QsCompiler.SyntaxTokens;
 using Microsoft.Quantum.QsCompiler.SyntaxTree;
 using Microsoft.Quantum.QsCompiler.Transformations.Core;
 
 namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
 {
+    using Concretion = HashSet<(QsTypeParameter, ResolvedType)>;
+
+    internal struct Request
```

Comment:
Is there anywhere where you actually set the properties of a Request and a Response below? What's the reason for the struct?
Otherwise I might lean towards having a class with readonly properties.

Reply:
The reason for the structs is that structs are generally more light-weight than classes from an efficiency perspective, so I was hoping here to eliminate some overhead with these, which are little more than tuples with named fields. Their members are never assigned to after initialization, so they are treated as immutable, but I can make some changes to reinforce that immutability.

Reply:
Yes, that might be good. I'm not sure if there are universally applicable rules of thumb for efficiency. I tend to like immutable data structures but that is definitively somewhat rooted in my preference for a functional setup if such a setup makes sense.